### PR TITLE
Add micro-ROS reconnections and liveliness tutorial 

### PR DESCRIPTION
**Pull request by Acuadros95**

### CODE REVIEW

File: docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+---------------------------------
+
+The main problem with the previous section method is that entities destruction may only happen on the client side. This implies that other ROS2 entities will not be aware of the micro-ROS client destruction.
+
+The Hard liveliness check mechanism allows the micro-ROS agent to ping the client periodically. This way, the agent will take care of ensuring that the micro-ROS client is alive and will destroy the created entities if a certain timeout happens without any response from the client side. This means that the nodes, publishers, subscribers (and any other entity) created by the client will be removed from the ROS 2 graph.
+
+Note that the client shall also be aware of the disconnection to create the micro-ROS entities again, this can be achieved by including the previous section approach.
+
+.. note::
+    The micro-ROS client shall shall spin an executor to give a response to the agent ping messages, a empty executor can be used for this purpose.
```
[the remark this answers is not in the document]
Comment (pablogs9) [reply docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst:77]: Also mention that this hard liveliness approach is intelligent in the sense of not sending additional data (ping messages) if the client and the agent are communicating normally. 

In general, we shall keep clear that this feature will **only use the unused bandwidth so it does not have penalty on the througput**

File: docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+Configuration
+^^^^^^^^^^^^^
+
+This feature is enabled by default in the micro-ROS Agent and must be enabled by means of a CMake argument in the micro-ROS client:
+
+    - ``UCLIENT_HARD_LIVELINESS_CHECK``: Enable hard liveliness check
+    - ``UCLIENT_HARD_LIVELINESS_CHECK_TIMEOUT``: Configure connection timeout in milliseconds (Default value: 10000).
```
Comment (pablogs9) [inline docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst:85]: Add here something like "the colcon.meta" shall looks like:

```
         "microxrcedds_client": {
            "cmake-args": [
                "-DUCLIENT_HARD_LIVELINESS_CHECKG=ON",
                "-DUCLIENT_HARD_LIVELINESS_CHECK_TIMEOUT=5000"
            ]
        },
```

File: docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+This configured timeout will be passed to the agent once the micro-ROS session is created.
```
Comment (pablogs9) [inline docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst:87]: Maybe this sentence is redundant

File: docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+Note that the client shall also be aware of the disconnection to create the micro-ROS entities again, this can be achieved by including the previous section approach.
+
+.. note::
+    The micro-ROS client shall shall spin an executor to give a response to the agent ping messages, a empty executor can be used for this purpose.
+
+Configuration
+^^^^^^^^^^^^^
+
+This feature is enabled by default in the micro-ROS Agent and must be enabled by means of a CMake argument in the micro-ROS client:
```
Comment (pablogs9) [inline docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst:82]: "must be enabled" in bold

**Review by pablogs9 — CHANGES_REQUESTED** [review]

File: docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+This includes temporal transport faults, agent or client restarts, client plug and play, etc.
+
+Client side: Ping API
+---------------------
+
+Reconnections can be handle manually on the client side using the agent ping functionality with the following sequence:
+
+1. Wait until agent is reachable.
+2. Create needed micro-ROS entities.
+3. Use the entities as usual: Spin, publish, etc.
```
[the remark this answers is not in the document]
Comment (Acuadros95) [reply docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst]: https://github.com/eProsima/vulcanexus/pull/85/commits/c84c9f3450450c108bc69692379e584a85abc4d7

File: docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+This includes temporal transport faults, agent or client restarts, client plug and play, etc.
+
+Client side: Ping API
+---------------------
+
+Reconnections can be handle manually on the client side using the agent ping functionality with the following sequence:
+
+1. Wait until agent is reachable.
+2. Create needed micro-ROS entities.
```
[the remark this answers is not in the document]
Comment (Acuadros95) [reply docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst]: https://github.com/eProsima/vulcanexus/pull/85/commits/c84c9f3450450c108bc69692379e584a85abc4d7

File: docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+A full example can be found on micro-ROS for Arduino repository `micro-ros_reconnection <https://github.com/micro-ROS/micro_ros_arduino/blob/humble/examples/micro-ros_reconnection_example/micro-ros_reconnection_example.ino>`_ example.
+
+.. note::
+    Details on the Ping API usage can be found in the :ref:`Middleware API: Ping agent <tutorials_micro_user_middleware_ping>` tutorial.
+
+Agent side: Hard liveliness check
+---------------------------------
+
+The main problem with the previous section method is that entities destruction may only happen on the client side. This implies that other ROS2 entities will not be aware of the micro-ROS client destruction.
```
[the remark this answers is not in the document]
Comment (Acuadros95) [reply docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst]: https://github.com/eProsima/vulcanexus/pull/85/commits/c84c9f3450450c108bc69692379e584a85abc4d7

[the remark this answers is not in the document]
Comment (Acuadros95) [reply docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst]: :cry: https://github.com/eProsima/vulcanexus/pull/85/commits/c84c9f3450450c108bc69692379e584a85abc4d7

File: docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (Acuadros95) [reply docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst]: https://github.com/eProsima/vulcanexus/pull/85/commits/c84c9f3450450c108bc69692379e584a85abc4d7

File: docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst (hunk 2 above)
Comment (Acuadros95) [reply docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst:85]: https://github.com/eProsima/vulcanexus/pull/85/commits/c84c9f3450450c108bc69692379e584a85abc4d7

File: docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst (hunk 4 above)
Comment (Acuadros95) [reply docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst]: https://github.com/eProsima/vulcanexus/pull/85/commits/c84c9f3450450c108bc69692379e584a85abc4d7

File: docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst (hunk 3 above)
Comment (Acuadros95) [reply docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst]: Removed on https://github.com/eProsima/vulcanexus/pull/85/commits/c84c9f3450450c108bc69692379e584a85abc4d7

File: docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        }
+
+        // Create micro-ROS entities
+        create_entities();
+
+        // Spin period
+        const unsigned int spin_timeout = RCL_MS_TO_NS(100);
+
+        // Spin until connection fails
+        while(RMW_RET_OK != rclc_executor_spin_some(&executor, spin_timeout))
```
[the remark this answers is not in the document]
Comment (Acuadros95) [reply docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst]: Updated code here https://github.com/eProsima/vulcanexus/pull/85/commits/1e1ecc339c80762ff57799f6527f25db683f55bd

File: docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+Reconnections and liveliness
+============================
+
+.. contents:: Table of Contents
+    :depth: 1
+    :local:
+    :backlinks: none
+
+This tutorial aims at providing insight on mechanism for handling reconnections between the micro-ROS agent and the client and assert the connection liveliness.
```
[the remark this answers is not in the document]
Comment (Acuadros95) [reply docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst:11]: Maybe bold is too much as this terms are used a lot in each tutorial

[the remark this answers is not in the document]
Comment (Acuadros95) [reply docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst]: Capitalized both terms here: https://github.com/eProsima/vulcanexus/pull/85/commits/0c49d5dbc62718594c5b476501d1f8b7825c6039

I would open other PR to fix other tutorials or references on other files.

File: docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+.. note::
+    Details on the Ping API usage can be found in the :ref:`Middleware API: Ping agent <tutorials_micro_user_middleware_ping>` tutorial.
+
+Agent side: Hard liveliness check
+---------------------------------
+
+The main problem with the previous section method is that entities destruction may only happen on the client side. This implies that other ROS2 entities will not be aware of the micro-ROS client destruction.
+
+The Hard liveliness check mechanism allows the micro-ROS agent to ping the client periodically. This way, the agent will take care of ensuring that the micro-ROS client is alive and will destroy the created entities if a certain timeout happens without any response from the client side. This means that the nodes, publishers, subscribers (and any other entity) created by the client will be removed from the ROS 2 graph.
```
[the remark this answers is not in the document]
Comment (Acuadros95) [reply docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst]: https://github.com/eProsima/vulcanexus/pull/85/commits/0c49d5dbc62718594c5b476501d1f8b7825c6039

File: docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (Acuadros95) [reply docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst]: Added a explanation here https://github.com/eProsima/vulcanexus/pull/85/commits/bdca02da60b1f1fe4ccd1a4eaa0883a63c96b51d

File: docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+.. contents:: Table of Contents
+    :depth: 1
+    :local:
+    :backlinks: none
+
+This tutorial aims at providing insight on mechanism for handling reconnections between the micro-ROS agent and the client and assert the connection liveliness.
+
+This includes temporal transport faults, agent or client restarts, client plug and play, etc.
+
+Client side: Ping API
```
[the remark this answers is not in the document]
Comment (Acuadros95) [reply docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst:19]: Modified here https://github.com/eProsima/vulcanexus/pull/85/commits/f84cee29b14d479b40ec91cdbd3cc77dc6170b8d

File: docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+============================
+
+.. contents:: Table of Contents
+    :depth: 1
+    :local:
+    :backlinks: none
+
+This tutorial aims at providing insight on mechanism for handling reconnections between the micro-ROS agent and the client and assert the connection liveliness.
+
+This includes temporal transport faults, agent or client restarts, client plug and play, etc.
```
[the remark this answers is not in the document]
Comment (Acuadros95) [reply docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst]: Improved here https://github.com/eProsima/vulcanexus/pull/85/commits/b3ba8eb9cf5948bcc409433228e6f636967c62e2

File: docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+
+State machine approach
+^^^^^^^^^^^^^^^^^^^^^^
+
+Reconnections can be handle manually on the micro-ROS Client side using the Agent ping functionality with the following sequence:
+
+1. Wait until Agent is reachable.
+2. Create required micro-ROS entities.
+3. Use the micro-ROS API as usual: spin entities, publish, etc.
+4. When a failure is detected in steps 2 or 3, destroy the created entities and go back to the first step.
```
Comment (dcoello-eProsima) [inline docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst:34]: is there an easy way to add sequence or activity diagrams to rst?? like mermaid on githuib markdown, I feel that an activity diagram for this sequence would be nice but only if its easy to include

**Review by dcoello-eProsima — APPROVED** [review]

File: docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst (hunk 13 above)
Comment (pablogs9) [reply docs/rst/tutorials/micro/handle_reconnections/handle_reconnections.rst:34]: Good idea, maybe another PR

**Review by pablogs9 — APPROVED** [review]

**Pull request merged**